runtime/binding-filesystem/src/main/java/io/aklivity/zilla/runtime/binding/filesystem/internal/stream/FileSystemServerFactory.java

```diff
@@ -283,25 +330,113 @@ private void onAppBegin(
             state = FileSystemState.openingInitial(state);
 
             doAppWindow(traceId);
+            String currentTag = calculateTag();
+            if (tag == null || tag.isEmpty() || !tag.equals(currentTag))
+            {
+                Flyweight replyBeginEx = getReplyBeginEx(currentTag);
+                doAppBegin(traceId, replyBeginEx);
+                flushAppData(traceId);
+            }
+            else
+            {
+                long timeoutAt = now().toEpochMilli() + beginEx.timeout();
+                long timeoutId = signaler.signalAt(timeoutAt, routeId, replyId, TIMEOUT_EXPIRED_SIGNAL_ID, 0);
+                watchedFile = new FileSystemWatcher.WatchedFile(
+                    resolvedPath, symlinks, this::calculateTag, tag, timeoutId, routeId, replyId);
+                fileSystemWatcher.watch(watchedFile);
+            }
+        }
 
-            final int capabilities = beginEx.capabilities();
-            final String16FW path = beginEx.path();
-            final long size = attributes.size();
-            final long modifiedTime = attributes.lastModifiedTime().toMillis();
+        private Flyweight getReplyBeginEx(String newTag)
+        {
+            int capabilities = tag == null || !tag.equals(newTag) ? this.capabilities : 0;
+            return getReplyBeginEx(newTag, capabilities);
+        }
 
-            Flyweight newBeginEx = beginExRW
+        private Flyweight getReplyBeginEx(
+            String newTag,
+            int capabilities)
+        {
+            attributes = getAttributes();
+            long size = 0L;
+            if (attributes != null)
+            {
+                size = attributes.size();
+            }
+            Flyweight replyBeginEx = beginExRW
                 .wrap(extBuffer, 0, extBuffer.capacity())
                 .typeId(fileSystemTypeId)
                 .capabilities(capabilities)
-                .path(path)
+                .path(relativePath)
                 .type(type)
                 .payloadSize(size)
-                .modifiedTime(modifiedTime)
+                .tag(newTag)
                 .build();
+            return replyBeginEx;
+        }
```

Comment:
We typically don't code building of flyweights like this, as return values for a method, because they must be only used on stack just before the call to write them out before they are available for reuse.

Recommend moving this code into `doAppBegin(tag)` instead, perhaps with an overload of `doAppBegin(tag, capabilities)`.